prow/plugins/trigger/trigger.go

```diff
@@ -153,3 +156,97 @@ func isUserTrusted(ghc githubClient, user, trustedOrg, org string) (bool, error)
 	}
 	return orgMember, nil
 }
+
+func runOrSkipRequested(c client, pr *github.PullRequest, requestedJobs []config.Presubmit, forceRunContexts map[string]bool, body, eventGUID string) error {
+	org := pr.Base.Repo.Owner.Login
+	repo := pr.Base.Repo.Name
+	number := pr.Number
+
+	baseSHA, err := c.GitHubClient.GetRef(org, repo, "heads/"+pr.Base.Ref)
+	if err != nil {
+		return err
+	}
+
+	getChanges := fileChangesGetter(c.GitHubClient, org, repo, number)
+	// shouldRun indicates if a job should actually run.
+	shouldRun := func(j config.Presubmit) (bool, error) {
+		if !j.RunsAgainstBranch(pr.Base.Ref) {
+			return false, nil
+		}
+		if j.RunIfChanged == "" || forceRunContexts[j.Context] || j.TriggerMatches(body) {
+			return true, nil
+		}
+		changes, err := getChanges()
+		if err != nil {
+			return false, err
+		}
+		return j.RunsAgainstChanges(changes), nil
+	}
+
+	// For each job determine if any sharded version of the job runs.
+	// This in turn determines which jobs to run and which contexts to mark as "Skipped".
+	//
+	// Note: Job sharding is achieved with presubmit configurations that overlap on
+	// name, but run under disjoint circumstances. For example, a job 'foo' can be
+	// sharded to have different pod specs for different branches by
+	// creating 2 presubmit configurations with the name foo, but different pod
+	// specs, and specifying different branches for each job.
+	var toRunJobs []config.Presubmit
+	toRun := sets.NewString()
+	toSkip := sets.NewString()
+	for _, job := range requestedJobs {
+		runs, err := shouldRun(job)
+		if err != nil {
+			return err
```

Comment:
So if we fail to list the changes once, we run no jobs at all? Is that a regression from current behavior?

Reply:
We only try to list the changes once, but if that fails we don't run any jobs. This is the same as the current behavior. We could switch to a best-effort approach and still trigger non run_if_changed jobs in this case, but I'm not sure if that is really much better. It could be more confusing to user to see some jobs start and others not? I don't feel strongly either way.

Reply:
we should ensure that the error is logged properly either way, and potentially add retries to github read operations as necessary -- but I don't think there's much we can do if github gives persistent 500s for an API

Reply:
The error will be logged by hook after the trigger plugin returns it and the github client already does retries.